### PR TITLE
Added some custom exceptions to be used in services. Added @Override …

### PR DESCRIPTION
**Pull request by ivanlavrinenko**
…of equals method in StorageImpl.

Tested handlers and services.

### CODE REVIEW

**Review by olekskov — CHANGES_REQUESTED** [review]
Build is failed. Please recheck your test cases
Probably problem is in a way you write path, like
String path = "src\\test\\resources\\report.csv";
I don't think it should work with double \\

**Comment by ivanlavrinenko** [comment]
> Build is failed. Please recheck your test cases Probably problem is in a way you write path, like String path = "src\ \test\ \resources\ \report.csv"; I don't think it should work with double \ \

i have build succes in my project, i've also tried to deal with single \ and it gives a compiler error

File: src/test/java/core/basesyntax/BalanceOperationHandlerTest.java [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    @Test
+    public void process_validData_Ok() {
+        final FruitTransaction fruitTransaction = new FruitTransaction(
+                FruitTransaction.Operation.BALANCE,
+                "orange",
+                888);
+        Storage expectedStorage = new StorageImpl();
+        expectedStorage.put("banana", 234);
+        expectedStorage.put("apple", 327);
+        expectedStorage.put("orange", 888);
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/BalanceOperationHandlerTest.java:39]: Save this to constant and use them

File: src/test/java/core/basesyntax/CsvFileWriterServiceTest.java [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Test
+    public void writeToFile_invalidPath_NotOk() {
+        String validData = """
+                fruit,quantity
+                banana,123
+                apple,909
+                coconut,90""";
+        String path = "shos/neponiatne";
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/CsvFileWriterServiceTest.java:28]: )))

File: src/test/java/core/basesyntax/CsvFileWriterServiceTest.java [hunk 3]
```diff
@@ -0,0 +1,71 @@
...
+    }
+
+    @Test
+    public void writeToFile_validPath_Ok() {
+        String validData = """
+                fruit,quantity
+                banana,123
+                apple,909
+                coconut,90""";
+        String path = "src/test/resources/report.csv";
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/CsvFileWriterServiceTest.java:41]: Path can save to constant

File: src/test/java/core/basesyntax/PurchaseOperationHandlerTest.java [hunk 4]
```diff
@@ -0,0 +1,42 @@
...
+    @BeforeAll
+    public static void setPurchaseOperationHandler() {
+        purchaseOperationHandler = new PurchaseOperationHandler();
+    }
+
+    @BeforeEach
+    public void setStorage() {
+        storage = new StorageImpl();
+        storage.put("banana", 234);
+        storage.put("apple", 327);
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/PurchaseOperationHandlerTest.java:27]: Constants

File: src/test/java/core/basesyntax/ReturnOperationHandlerTest.java [hunk 5]
```diff
@@ -0,0 +1,42 @@
...
+    @BeforeAll
+    public static void setReturnOperationHandler() {
+        returnOperationHandler = new ReturnOperationHandler();
+    }
+
+    @BeforeEach
+    public void setStorage() {
+        storage = new StorageImpl();
+        storage.put("banana", 234);
+        storage.put("apple", 327);
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/ReturnOperationHandlerTest.java:27]: Fix it for all

**Review by Sanderfuck — CHANGES_REQUESTED** [review]
Make mini refactor of constants

File: src/test/java/core/basesyntax/ReportServiceTest.java [hunk 6]
```diff
@@ -0,0 +1,38 @@
...
+
+public class ReportServiceTest {
+    private static final String BANANA = "banana";
+    private static final String APPLE = "apple";
+    private static final String COCONUT = "coconut";
+    private static ReportService reportService;
+    private static Storage storage;
+
+    @BeforeAll
+    public static void setReportService() {
```
Comment (okuzan) [inline src/test/java/core/basesyntax/ReportServiceTest.java:21]:
```suggestion
    public static void init() {
```

File: src/test/java/core/basesyntax/ReportServiceTest.java [hunk 7]
```diff
@@ -0,0 +1,38 @@
...
+        reportService = new ReportServiceImpl();
+    }
+
+    @BeforeEach
+    public void setStorage() {
```
Comment (okuzan) [inline src/test/java/core/basesyntax/ReportServiceTest.java:26]:
```suggestion
    public void setup() {
```

File: src/test/java/core/basesyntax/ReportServiceTest.java [hunk 8]
```diff
@@ -0,0 +1,38 @@
...
+        storage = new StorageImpl();
+    }
+
+    @Test
+    public void createReport_validData() {
```
Comment (okuzan) [inline src/test/java/core/basesyntax/ReportServiceTest.java:31]:
```suggestion
    public void createReport_validData_ok() {
```

File: src/test/java/core/basesyntax/PurchaseOperationHandlerTest.java [hunk 9]
```diff
@@ -0,0 +1,44 @@
...
+
+    @BeforeEach
+    public void setStorage() {
+        storage = new StorageImpl();
+        storage.put(BANANA, 234);
+        storage.put(APPLE, 327);
+    }
+
+    @Test
+    public void process_validData_Ok() {
```
Comment (okuzan) [inline src/test/java/core/basesyntax/PurchaseOperationHandlerTest.java:33]:
```suggestion
    public void process_validData_ok() {
```

File: src/test/java/core/basesyntax/ReportServiceTest.java (hunk 6 above)
Comment (okuzan) [reply src/test/java/core/basesyntax/ReportServiceTest.java:21]: unusual naming, fix everywhere

File: src/test/java/core/basesyntax/PurchaseOperationHandlerTest.java (hunk 9 above)
Comment (okuzan) [reply src/test/java/core/basesyntax/PurchaseOperationHandlerTest.java:33]: let's check what happens if we try to purchase amount bigger than what we have in a storage. In other handlers, what happens if we pass a negative amount? FYI If you need to make changes to the code in /main, do it

File: src/test/java/core/basesyntax/ConversionServiceTest.java [hunk 10]
```diff
@@ -0,0 +1,52 @@
...
+        List<String[]> invalidList = List.of(
+                new String[]{"b", "banana", "qwerty"},
+                new String[]{"p", "apple", "23"});
+        assertThrows(NonNumericDataException.class,
+                () -> conversionService.convert(invalidList),
+                "Non-numeric data provided as quantity!");
+    }
+
+    @Test
+    public void convert_validData_Ok() {
```
Comment (okuzan) [inline src/test/java/core/basesyntax/ConversionServiceTest.java:43]: let's user lowercase after underscore everywhere

File: src/test/java/core/basesyntax/ConversionServiceTest.java [hunk 11]
```diff
@@ -0,0 +1,52 @@
...
+        List<String[]> validList = List.of(
```
Comment (okuzan) [inline src/test/java/core/basesyntax/ConversionServiceTest.java:44]: we usually don't include data type in variable's name, Java is a strongly typed language, better call it validTransactions

File: src/test/java/core/basesyntax/ConversionServiceTest.java [hunk 12]
```diff
@@ -0,0 +1,52 @@
...
+                new String[]{"b", "banana", "23"},
+                new String[]{"p", "apple", "234"});
+        assertDoesNotThrow(() -> conversionService.convert(validList),
+                "Provided a non-existing operation code!");
+        assertDoesNotThrow(() -> conversionService.convert(validList),
+                "Non-numeric data provided as quantity!");
```
Comment (okuzan) [inline src/test/java/core/basesyntax/ConversionServiceTest.java:50]: having two same assertions on same method call is redundant

File: src/test/java/core/basesyntax/OperationStrategyTest.java [hunk 13]
```diff
@@ -0,0 +1,58 @@
...
+public class OperationStrategyTest {
+    private static OperationStrategy operationStrategy;
+
+    @BeforeAll
+    public static void setUp() {
+        Map<FruitTransaction.Operation, OperationHandler> operationHandlerMap = new HashMap<>();
+        operationHandlerMap.put(FruitTransaction.Operation.BALANCE,
+                new BalanceOperationHandler());
+        operationHandlerMap.put(FruitTransaction.Operation.RETURN,
+                new ReturnOperationHandler());
```
Comment (okuzan) [inline src/test/java/core/basesyntax/OperationStrategyTest.java:27]: you can use Map.of

**Review by okuzan — CHANGES_REQUESTED** [review]
Separate test classes into the same folder hierarchy as in `main` package

File: src/test/java/core/basesyntax/strategy/handler/BalanceOperationHandlerTest.java [hunk 14]
```diff
@@ -0,0 +1,44 @@
...
+import core.basesyntax.db.Storage;
+import core.basesyntax.db.StorageImpl;
+import core.basesyntax.model.FruitTransaction;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class BalanceOperationHandlerTest {
+    private static final String BANANA = "banana";
+    private static final String APPLE = "apple";
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/strategy/handler/BalanceOperationHandlerTest.java:14]: You don`t need to create same constants if you already have, lets reuse

File: src/test/java/core/basesyntax/strategy/handler/PurchaseOperationHandlerTest.java [hunk 15]
```diff
@@ -0,0 +1,42 @@
...
+import core.basesyntax.db.Storage;
+import core.basesyntax.db.StorageImpl;
+import core.basesyntax.model.FruitTransaction;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class PurchaseOperationHandlerTest {
+    private static final String BANANA = "banana";
+    private static final String APPLE = "apple";
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/strategy/handler/PurchaseOperationHandlerTest.java:14]: The same

File: src/test/java/core/basesyntax/strategy/handler/SupplyOperationHandlerTest.java [hunk 16]
```diff
@@ -0,0 +1,42 @@
...
+import core.basesyntax.db.Storage;
+import core.basesyntax.db.StorageImpl;
+import core.basesyntax.model.FruitTransaction;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class SupplyOperationHandlerTest {
+    private static final String BANANA = "banana";
+    private static final String APPLE = "apple";
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/strategy/handler/SupplyOperationHandlerTest.java:14]:
```suggestion
 
```

File: src/test/java/core/basesyntax/strategy/handler/ReturnOperationHandlerTest.java [hunk 17]
```diff
@@ -0,0 +1,42 @@
...
+import core.basesyntax.db.Storage;
+import core.basesyntax.db.StorageImpl;
+import core.basesyntax.model.FruitTransaction;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+public class ReturnOperationHandlerTest {
+    private static final String BANANA = "banana";
+    private static final String APPLE = "apple";
```
Comment (Sanderfuck) [inline src/test/java/core/basesyntax/strategy/handler/ReturnOperationHandlerTest.java:14]:
```suggestion
```

**Review by Sanderfuck — CHANGES_REQUESTED** [review]

**Review by kozhukhovsky — APPROVED** [review]

**Pull request closed without merging**